wordpress-mobile/WordPress-Android
# [Home Page Picker] Adds polling mechanism for not generated thumbnails

**Supplemental fix for:** https://github.com/wordpress-mobile/gutenberg-mobile/issues/2718

**Depends on:**
- https://github.com/wordpress-mobile/WordPress-FluxC-Android/pull/1771

## Description

This PR implements a custom Glide loader that prevents following redirects and applies a retry policy with a specific number of retries (10).

This is needed because the mshot backend service redirects to a loading gif image when the thumbnail is not ready. This occurs when the thumbnail has not been accessed recently for the specific language and viewport size. When this thumbnail is requested it is cached on the server and is readily available according to the server cache policy.

## To test:

<details>
<summary><strong>To disable or enable the development version of Home Page Picker</strong></summary>

- Open the app from the build that allows FeatureFlags such as a PR build or a local development build
    - By default, the modal layout picker is enabled only in development builds.
    - From the site page:
        1. Click on your Gravatar.
        1. Click on App Settings.
        1. Click on Test feature configuration.
        1. Toggle "HomePagePickerFeatureConfig" to enable or disable the picker
        1. Restart the App
</details>

<details>
<summary><strong>To start the Home Page Picker screen</strong></summary>

1. Start the site creation flow (e.g. *Choose site > Add button*)
1. Select the *Create WordPress.com site* option
</details>

1. Reinstall the app or clean the app cache (Settings > Apps > WordPress > Storage & cache > Clear)
1. Login
1. Set the application language to one that has no cached thumbnails on server (try `ja` or `ko`)
1. Start the Home Page Picker 
1. Expect to see the ghost images for a short time then see the images load

*Note: an alternative way to test this is to tweak the thumbnail url to request a different viewport [programmatically](https://github.com/wordpress-mobile/WordPress-Android/pull/13386/commits/d29437dd40ce928262465e9daa33b92c9c15bfcf)*

PR submission checklist:
